### ISSUE TITLE
Implement High-Quality Voice Cloning with XTTS-v2 Fine-Tuning

### ISSUE THREAD

**Issue by JunlinYuan** [JunlinYuan/tts#1]
# Implement High-Quality Voice Cloning with XTTS-v2 Fine-Tuning

## 🎯 Overview

Currently, the TTS MCP server only supports zero-shot voice cloning ("low" quality), which provides basic voice similarity using 6-30 second samples. This enhancement will implement fine-tuning capabilities for XTTS-v2 to achieve significantly higher quality voice cloning through custom model training.

### Fine-Tuning Best Practices
- **Training Epochs**: Complex voices require 100+ epochs for adequate training
- **Optimizer**: AdamW is recommended for stable results with minimal tuning
- **Batch Size**: 8-32 with gradient accumulation for memory efficiency
- **Hardware**: 8GB+ VRAM minimum, optimizable for M1 Max with CPU fallback
- **Data Format**: LJSpeech format with proper metadata.csv structure

### Quality Improvements
- Fine-tuning significantly improves voice similarity and accent accuracy beyond zero-shot
- Custom models capture speaker-specific characteristics more effectively
- Training on 30+ minutes of clean audio produces near-human quality results

### Technical Architecture
XTTS-v2 uses a GPT2 model with Perceiver-based speaker conditioning, enabling multiple reference support and improved voice consistency across languages.

### Existing Architecture
- ✅ Zero-shot generation via `TTSEngine.generate_zero_shot()`
- ✅ Modular design with separate voice and audio managers
- ✅ MCP server with 4 tools
- ✅ Quality parameter accepted but not implemented

### Implementation Gaps
- ❌ No fine-tuning training pipeline
- ❌ No model management for custom trained models  
- ❌ No quality-based routing (`generate_tts` always uses zero-shot)
- ❌ No training data preprocessing tools
- ❌ Empty `models/` directory structure

### 1. Fine-Tuning Engine (`mcp_server/training_engine.py`)
```python
class TrainingEngine:
    def start_fine_tuning(self, identity: str, dataset_path: str, config: dict) -> str:
        # Initialize fine-tuning process with progress tracking
        
    def get_training_status(self, job_id: str) -> dict:
        # Return training progress, loss metrics, ETA
        
    def stop_training(self, job_id: str) -> bool:
        # Gracefully stop training process
```

### 2. Dataset Preparation Tools
- Audio preprocessing and validation
- LJSpeech format conversion
- Metadata.csv generation
- Audio quality assessment and filtering

### 3. Model Management System
- Custom model storage in `voice_identities/{identity}/models/`
- Model versioning and metadata tracking
- Quality-based model selection logic
- Model validation and testing

### 4. Enhanced TTS Engine
```python
def generate_fine_tuned(self, text: str, model_path: str, language: str) -> bytes:
    # Generate using custom fine-tuned model
    
def select_model_by_quality(self, identity: str, quality: str) -> str:
    # Route to appropriate model based on quality parameter
```

### 5. New MCP Tools
- `start_fine_tuning`: Initialize training process
- `get_training_status`: Monitor training progress
- `prepare_dataset`: Preprocess training data
- `validate_model`: Test model quality

### Core Functionality
- [ ] Fine-tuning pipeline successfully trains XTTS-v2 models
- [ ] `generate_tts` with `quality="high"` uses fine-tuned models
- [ ] Training progress is trackable and interruptible
- [ ] Models are saved and loaded correctly from `voice_identities/{identity}/models/`

### Performance Targets
- [ ] Training completes within 2-4 hours for 30-minute datasets on M1 Max
- [ ] Fine-tuned models show measurable quality improvement over zero-shot
- [ ] Memory usage stays within 16GB during training
- [ ] Generation speed remains <5 seconds per sentence

### Integration Requirements
- [ ] Maintains compatibility with existing MCP server architecture
- [ ] Follows established error handling patterns
- [ ] Preserves existing zero-shot functionality
- [ ] Integrates with current voice identity management

### Quality Metrics
- [ ] Voice similarity scores improve by 20%+ over zero-shot
- [ ] Accent accuracy significantly enhanced
- [ ] Consistent quality across English and French
- [ ] Model robustness validated with diverse text inputs

### Unit Tests
- Training pipeline components
- Model saving/loading functionality
- Dataset preprocessing validation
- Quality routing logic

### Integration Tests
- End-to-end fine-tuning workflow
- MCP tool interactions
- Model switching between quality levels
- Error handling during training

### Performance Tests
- Training speed benchmarks
- Memory usage profiling
- Audio quality assessments
- Model inference performance

### Hardware Optimization
- Leverage M1 Max unified memory architecture
- Implement gradient accumulation for memory efficiency
- CPU fallback for lower-end hardware
- Progress checkpointing for long training sessions

### Data Management
- Secure handling of voice training data
- Efficient storage of multiple model versions
- Cleanup utilities for old models
- Privacy considerations for voice data

### Error Handling
- Clear training failure diagnostics
- Recovery from interrupted training
- Model corruption detection
- Insufficient data warnings

### Phase 1: Core Training Infrastructure
- Implement `TrainingEngine` class
- Add dataset preprocessing tools
- Create model management system

### Phase 2: MCP Integration
- Add new MCP tools for training
- Implement progress tracking
- Integrate with existing server

### Phase 3: Quality Routing
- Enhance `TTSEngine` with model selection
- Implement quality-based generation
- Add model validation

### Phase 4: Testing & Optimization
- Comprehensive testing suite
- Performance optimization
- Documentation and examples

## 🎯 Success Metrics

- High-quality voice cloning demonstrably superior to zero-shot
- Training pipeline robust and user-friendly
- Seamless integration with existing MCP architecture
- Performance targets met on M1 Max hardware
- Clear path to Swift frontend integration (Phase 6 roadmap)

## 🔗 Related Issues

This implements the planned "Phase 5: Fine-Tuning Support" from the project roadmap, enabling the transition from MVP to production-quality voice cloning capabilities.

---

**Priority**: High
**Complexity**: High
**Impact**: High - Transforms system from proof-of-concept to production-ready
